tests/cpp/test_sdpa_node.cpp

```diff
@@ -232,4 +229,268 @@ TEST_F(SDPATest, PairwiseRootDomainMap) {
+  auto tv_seed = makeConcreteTensor({}, DataType::Int);
+  tv_seed->setCpuScalar(true);
+  auto tv_offset = makeConcreteTensor({}, DataType::Int);
+  tv_offset->setCpuScalar(true);
+
+  fusion->addInput(tv_grad_output);
+  fusion->addInput(tvq);
+  fusion->addInput(tvk);
+  fusion->addInput(tvv);
+  fusion->addInput(tv_output);
+  fusion->addInput(tv_logsumexp);
+  fusion->addInput(tv_cumq);
+  fusion->addInput(tv_cumk);
+  fusion->addInput(tv_seed);
+  fusion->addInput(tv_offset);
+
+  auto tvgrad = sdpfa_bwd(
+      tv_grad_output,
+      tvq,
+      tvk,
+      tvv,
+      tv_output,
+      tv_logsumexp,
+      tv_cumq,
+      tv_cumk,
+      /*max_q=*/IrBuilder::create<Val>(*query_seq_len.maybe_as_int()),
+      /*max_k=*/IrBuilder::create<Val>(*key_seq_len.maybe_as_int()),
+      /*dropout_p=*/IrBuilder::create<Val>(0.0),
+      /*is_causal=*/IrBuilder::create<Val>(false),
+      tv_seed,
+      tv_offset,
+      /*scale=*/nullptr);
+
+  fusion->addOutput(tvgrad.grad_query);
+  fusion->addOutput(tvgrad.grad_key);
+  fusion->addOutput(tvgrad.grad_value);
+
+  at::Tensor grad_out = at::randn(attn_shape, options);
+
+  std::vector<c10::IValue> sdpa_bwd_inputs = {
+      grad_out,
+      q,
+      k,
+      v,
+      output,
+      log_sumexp,
+      cum_seq_q,
+      cum_seq_k,
+      philox_seed,
+      philox_offset};
+  FusionExecutor fe;
+  std::for_each(
+      fusion->outputs().begin(), fusion->outputs().end(), [&](Val* out) {
+        fusion->aliasOutputToInput(
+            out, /*input=*/nullptr, AllocationType::Evaluate);
+      });
+
+  fe.compileFusion(fusion.get(), sdpa_bwd_inputs);
+  auto out = fe.runFusion(sdpa_bwd_inputs);
+
+  auto [ref_grad_query, ref_grad_key, ref_grad_value] =
+      at::_scaled_dot_product_flash_attention_backward(
+          grad_out,
+          q,
+          k,
+          v,
+          output,
+          log_sumexp,
+          cum_seq_q,
+          cum_seq_k,
+          /*max_q=*/*query_seq_len.maybe_as_int(),
+          /*max_k=*/*key_seq_len.maybe_as_int(),
+          /*dropout_p=*/dropout_p,
+          /*is_causal=*/is_causal,
+          philox_seed,
+          philox_offset,
+          /*scale=*/scale);
+
+  testValidate(
+      fusion.get(),
+      out,
+      sdpa_bwd_inputs,
+      {ref_grad_query, ref_grad_key, ref_grad_value},
+      __LINE__,
+      __FILE__);
+}
+
+TEST_F(SDPATest, NonCausalAttnSymbolicBwd) {
+  NVFUSER_TEST_CUDA_ARCH_GUARD(8, 0);
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  std::vector<int64_t> q_shape({n, h, l, e});
+  std::vector<int64_t> k_shape({n, h, s, e});
+  std::vector<int64_t> v_shape({n, h, s, e});
+  std::vector<int64_t> attn_shape({n, h, l, e});
+
+  auto options = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+  at::Tensor q = at::randn(q_shape, options);
+  at::Tensor k = at::randn(k_shape, options);
+  at::Tensor v = at::randn(v_shape, options);
+
+  double dropout_p = 0.0;
```

Comment:
Might want to set this to a non-zero value to test that it's passed properly.

Reply:
since both reference and ExprEval uses the same torch implementation, if we start with the same random seed  we should still be able to compare the two.

On a separate note, I'm now wondering how are we testing nvfuser codegen's random behavior. Since we are not converging on the random number pattern with torch?

Reply:
We do compute the same random numbers as torch, with the exception that changing stride order in pytorch changes the generated values whereas our numbers always match the logical domain. But yes, in this case it is ATen vs ATen so it should be exactly the same.

Reply:
> We do compute the same random numbers as torch

Thanks a lot! I didn't know that.

I thought they matches by accident with small buffer and when the size goes up, pytorch matches it differently and would give a different pattern. That's probably some out-dated implementation. :)

Reply:
For reference: https://github.com/NVIDIA/Fuser/issues/557#issuecomment-1620466166

Reply:
We don't test that the numbers match anymore though, so this might no longer be the case and we should not rely on nvfuser computing the same random numbers.